src/Textarea.js

```diff
@@ -1,123 +1,111 @@
-import React, { Component } from 'react';
+import React, { cloneElement, useRef, useEffect } from 'react';
 import PropTypes from 'prop-types';
 import cx from 'classnames';
 import idgen from './idgen';
 import constants from './constants';
 
-class Textarea extends Component {
-  constructor(props) {
-    super(props);
-
-    this.id = props.id || idgen();
+const Textarea = ({
+  id,
+  children,
+  className,
+  s,
+  m,
+  l,
+  xl,
+  disabled,
+  noLayout,
+  placeholder,
+  icon,
+  label,
+  success,
+  error,
+  iconClassName,
+  onChange,
+  validate,
+  defaultValue,
+  value,
+  ...props
+}) => {
+  const _textarea = useRef(null);
+  const sizes = { s, m, l, xl };
+
+  let responsiveClasses;
+  if (!noLayout) {
+    responsiveClasses = { col: true };
+    constants.SIZES.forEach(size => {
+      responsiveClasses[size + sizes[size]] = sizes[size];
+    });
   }
 
-  componentDidMount() {
-    if (typeof M !== undefined) {
-      // eslint-disable-next-line react/prop-types
-      this.props['data-length'] && M.CharacterCounter.init(this._textarea);
-    }
-  }
-
-  componentDidUpdate(prevProps) {
-    const { value } = this.props;
-
-    if (typeof M !== 'undefined' && value !== prevProps.value) {
-      M.textareaAutoResize(this._textarea);
+  const wrapperClasses = cx('input-field', responsiveClasses);
+
+  const textareaProps = {
+    placeholder,
+    id,
+    value,
+    defaultValue,
+    disabled,
+    ...props
+  };
+
+  useEffect(() => {
+    // eslint-disable-next-line react/prop-types
+    if (props['data-length']) {
+      const instance = M.CharacterCounter.init(_textarea.current);
+
+      return () => {
+        instance && instance.destroy();
+      };
     }
-  }
-
-  render() {
-    const {
-      children,
-      className,
-      s,
-      m,
-      l,
-      xl,
-      disabled,
-      noLayout,
-      placeholder,
-      icon,
-      label,
-      success,
-      error,
-      iconClassName,
-      onChange,
-      validate,
-      defaultValue,
-      value,
-      ...other
-    } = this.props;
-
-    const sizes = { s, m, l, xl };
+  }, [_textarea, props]);
 
-    let responsiveClasses;
-    if (!noLayout) {
-      responsiveClasses = { col: true };
-      constants.SIZES.forEach(size => {
-        responsiveClasses[size + sizes[size]] = sizes[size];
-      });
+  useEffect(() => {
+    if (_textarea.current) {
+      M.textareaAutoResize(_textarea.current);
     }
+  }, [value]);
+
+  const renderLabel = () =>
```

Comment:
useMemo? we only want to recompute this if label has changed
`const renderLabel = useMemo(() => ...`

Reply:
Thing is, `renderLabel` depends on `[error, id, label, placeholder, success, value]`.
Moreover it might be better to [`useCallback`](https://it.reactjs.org/docs/hooks-reference.html#usecallback) instead of `useMemo`.

But anyway, [I'm not really sure if we should use them](https://kentcdodds.com/blog/usememo-and-usecallback).

If we really want to do that I suggest to implement these changes in a next release, prior benchmarking  them to see the performances gain, if any.

Let me know your thoughs :wink:

Reply:
Ah, ok, i've just read this. And i've been commenting in all the PR's to add `useMemo`. Ok let me have a thought and i'll get back to you tomorrow.